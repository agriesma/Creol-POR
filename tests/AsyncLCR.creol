
class Process(i: Init, u: Nat)
begin
  // unknown = 0 or null
  // chosen = 1
  // reported = 2
  var status: Nat
  var neighbor: Process

  op init == status := 0; i.getNeighbor(; neighbor); !neighbor.receive(u)

  op run == skip

  op receive(in v: Nat) ==
	if v > u
	then !neighbor.receive(v)
	else if v = u
	     then status := 1 fi fi
end

class Init
begin
  var p1 : Process
  var p2 : Process
  var p3 : Process

  op init == p1 := new Process(this, 1); p2 := new Process(this, 2);
    p3 := new Process(this, 3)

  op getNeighbor(out p: Process) ==
    if caller = p1 then p := p3 else
    if caller = p2 then p := p1 else p := p2 fi fi

end
