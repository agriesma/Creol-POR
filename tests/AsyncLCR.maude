load interpreter
mod PROGRAM is
pr INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< 'Process : Cl | Inh: noInh, Par: 'i ,, 'u, Att: 'i |-> null , 'u |-> null , 'status |-> null , 'neighbor |-> null, Mtds: 
  < 'init : Mtdname | Param: noAid, Latt: noSubst, Code: 'status ::= int(0) ; 'i . 'getNeighbor( emp : 'neighbor ) ; 'Dummy ! 'neighbor . 'receive ( 'u ) ; return ( emp ) > *
  < 'run : Mtdname | Param: noAid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < 'receive : Mtdname | Param: 'v, Latt: 'v |-> null, Code: if ( 'less[[ 'u # 'v ]] ) th 'Dummy ! 'neighbor . 'receive ( 'v ) el if ( 'equal[[ 'v # 'u ]] ) th 'status ::= int(1) el skip fi fi ; return ( emp ) >, Ocnt: 0 >

< 'Init : Cl | Inh: noInh, Par: noAid, Att: 'p1 |-> null , 'p2 |-> null , 'p3 |-> null, Mtds: 
  < 'init : Mtdname | Param: noAid, Latt: noSubst, Code: 'p1 ::= new 'Process( 'this # int(1) ) ; 'p2 ::= new 'Process( 'this # int(2) ) ; 'p3 ::= new 'Process( 'this # int(3) ) ; return ( emp ) > *
  < 'getNeighbor : Mtdname | Param: noAid, Latt: 'p |-> null, Code: if ( 'equal[[ 'caller # 'p1 ]] ) th 'p ::= 'p3 el if ( 'equal[[ 'caller # 'p2 ]] ) th 'p ::= 'p1 el 'p ::= 'p2 fi fi ; return ( 'p ) >, Ocnt: 0 >

.
endm
