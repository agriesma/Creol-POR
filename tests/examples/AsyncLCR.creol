/*
 * AsyncLCR.creol
 *
 * Simple solution to the asynchronous leader-election protocol.
 */

interface Process
begin
  with Process
    op receive (in v: Int)
end

class Process(i: Init, u: Int)
contracts Process
begin
  // unknown = 0 || null
  // chosen = 1
  // reported = 2
  var status: Int
  var neighbor: Process

  op init == status := 0; i.getNeighbor(; neighbor); !neighbor.receive(u)

  op run == skip

  with Process
    op receive(in v: Int) ==
	if v > u
	then !neighbor.receive(v)
	else if v = u
	     then status := 1 end end
end

interface Init
begin
  with Process
    op getNeighbor(out p: Process)
end

class Init
contracts Init
begin
  var p1 : Process
  var p2 : Process
  var p3 : Process

  op init == p1 := new Process(this, 1); p2 := new Process(this, 2);
    p3 := new Process(this, 3)

  with Process
    op getNeighbor(out p: Process) ==
      if caller = p1 then p := p3 else
      if caller = p2 then p := p1 else p := p2 end end

end
