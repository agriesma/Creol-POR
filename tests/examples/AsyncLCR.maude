load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Process" : Cl | Inh: noInh, Par: "i" , "u", Att: "i" |-> null , "u" |-> null , "status" |-> null , "neighbor" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null, Code: "status" ::= int(0) ; "label:0" ! "i" . "getNeighbor" ( emp ) ; ( "label:0" ? ( "neighbor" ) )  ; "label:1" ! "neighbor" . "receive" ( "u" ) ; free( "label:1" ) ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "receive" : Mtdname | Param: "v", Latt: "v" |-> null , "label:0" |-> null, Code: if ">" ( "v" # "u" ) th "label:0" ! "neighbor" . "receive" ( "v" ) ; free( "label:0" ) el if "=" ( "v" # "u" ) th "status" ::= int(1) el skip fi fi ; return ( emp ) >, Ocnt: 0 >

< "Init" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null , "p2" |-> null , "p3" |-> null, Mtds: 
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new "Process" ( "this" # int(1) ) ; "p2" ::= new "Process" ( "this" # int(2) ) ; "p3" ::= new "Process" ( "this" # int(3) ) ; return ( emp ) > *
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "p" |-> null, Code: if "=" ( "caller" # "p1" ) th "p" ::= "p3" el if "=" ( "caller" # "p2" ) th "p" ::= "p1" el "p" ::= "p2" fi fi ; return ( "p" ) >, Ocnt: 0 >

.
endm
