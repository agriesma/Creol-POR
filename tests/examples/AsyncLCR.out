==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Init" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null, "p3"
    |-> null, Mtds:
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "p" |-> null, Code: if "="(
    "caller" # "p1") th "p" ::= "p3" el if "="("caller" # "p2") th "p" ::= "p1"
    el "p" ::= "p2" fi fi ; return("p") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Process"("this" # int(1)) ; "p2" ::= new "Process"("this" # int(2)) ; "p3"
    ::= new "Process"("this" # int(3)) ; return(emp) >,
  Ocnt: 0 >

< "Process" : Cl | Inh: noInh, Par: "i", "u", Att: "i" |-> null, "neighbor" |->
    null, "status" |-> null, "u" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "status" ::= int(0) ; "label:0" ! "i" . "getNeighbor"(emp) ;
    "label:0" ?("neighbor") ; "label:1" ! "neighbor" . "receive"("u") ; free(
    "label:1") ; return(emp) > *
  < "receive" : Mtdname | Param: "v", Latt: "label:0" |-> null, "v" |-> null,
    Code: if ">"("v" # "u") th "label:0" ! "neighbor" . "receive"("v") ; free(
    "label:0") el if "="("v" # "u") th "status" ::= int(1) el skip fi fi ;
    return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >
==========================================
rewrite in PROGRAM : init
main("Init", emp) .
result Configuration: 
< ob("Init0") : Qu | Size: 10, Dealloc: noDealloc, Ev: noMsg >

< ob("Process0") : Qu | Size: 10, Dealloc: noDealloc, Ev: noMsg >

< ob("Process1") : Qu | Size: 10, Dealloc: noDealloc, Ev: noMsg >

< ob("Process2") : Qu | Size: 10, Dealloc: noDealloc, Ev: noMsg >

< ob("main") : Qu | Size: 1, Dealloc: noDealloc, Ev: noMsg >

< "Init" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null, "p3"
    |-> null, Mtds:
  < "getNeighbor" : Mtdname | Param: noVid, Latt: "p" |-> null, Code: if "="(
    "caller" # "p1") th "p" ::= "p3" el if "="("caller" # "p2") th "p" ::= "p1"
    el "p" ::= "p2" fi fi ; return("p") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Process"("this" # int(1)) ; "p2" ::= new "Process"("this" # int(2)) ; "p3"
    ::= new "Process"("this" # int(3)) ; return(emp) >,
  Ocnt: 1 >

< "Process" : Cl | Inh: noInh, Par: "i", "u", Att: "i" |-> null, "neighbor" |->
    null, "status" |-> null, "u" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: "status" ::= int(0) ; "label:0" ! "i" . "getNeighbor"(emp) ;
    "label:0" ?("neighbor") ; "label:1" ! "neighbor" . "receive"("u") ; free(
    "label:1") ; return(emp) > *
  < "receive" : Mtdname | Param: "v", Latt: "label:0" |-> null, "v" |-> null,
    Code: if ">"("v" # "u") th "label:0" ! "neighbor" . "receive"("v") ; free(
    "label:0") el if "="("v" # "u") th "status" ::= int(1) el skip fi fi ;
    return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 3 >

< ob("Init0") : "Init" | Att: "p1" |-> ob("Process0"), "p2" |-> ob("Process1"),
    "p3" |-> ob("Process2"), "this" |-> ob("Init0"),
  Pr: idle,
  PrQ: noProc, Lcnt: 3 >

< ob("Process0") : "Process" | Att: "i" |-> ob("Init0"), "neighbor" |-> ob(
    "Process2"), "status" |-> int(0), "this" |-> ob("Process0"), "u" |-> int(
    1),
  Pr: idle,
  PrQ: noProc, Lcnt: 7 >

< ob("Process1") : "Process" | Att: "i" |-> ob("Init0"), "neighbor" |-> ob(
    "Process0"), "status" |-> int(0), "this" |-> ob("Process1"), "u" |-> int(
    2),
  Pr: idle,
  PrQ: noProc, Lcnt: 6 >

< ob("Process2") : "Process" | Att: "i" |-> ob("Init0"), "neighbor" |-> ob(
    "Process1"), "status" |-> int(1), "this" |-> ob("Process2"), "u" |-> int(
    3),
  Pr: idle,
  PrQ: noProc, Lcnt: 5 >

< ob("main") : "NoClass" | Att: noSubst,
  Pr: idle,
  PrQ: noProc, Lcnt: 0 >
Bye.
